### PR TITLE
feat: Implemented query and comparison of previous/new SHA value

### PR DESCRIPTION
**Pull request by tmsavage**
### Files Edited
- app/graphql/types/query_type.rb
- app/javascript/components/app.jsx
- app/javascript/components/get-version.jsx

### Description
Implemented a new query to attain a current (dummy) SHA value and conditions to compare with a new (dummy) SHA value. If the two do not match, then there is an update to the application; the page will refresh and the recorded current SHA is overwritten by the new SHA value.

The query gets the value stored in a new variable in the .env file. Ex: `HEROKU_SLUG_COMMIT=foosomething`, which serves as the current dummy SHA value.

### CODE REVIEW

File: app/graphql/types/query_type.rb [hunk 1]
```diff
@@ -39,4 +39,14 @@ def primary_location
...
     end
   end
+  
+  #Define a field
+  field :schema_sha, String do
+    description "Get current sha"
+  end
+
+  def schema_sha
+    ENV['HEROKU_SLUG_COMMIT']
```
Comment (chewnoill) [inline app/graphql/types/query_type.rb:49]: I think we've switched this to `AppEnv`

see https://github.com/mojotech/helios2/pull/224

File: app/javascript/components/get-version.jsx [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+      if (error) {
+        return <ErrorMessage />;
+      }
+
+      var newValue = data.schemaSha;
+      if (lastValue == null) {
+        lastValue = newValue;
+      } else if (newValue !== lastValue) {
+        lastValue = newValue;
+        object.location.reload();
```
Comment (chewnoill) [inline app/javascript/components/get-version.jsx:30]: does this work?
I think you want `window.location.reload()`

**Review by chewnoill — COMMENTED** [review]
needs a squash and a rebase

File: app/javascript/components/get-version.jsx [hunk 3]
```diff
@@ -0,0 +1,38 @@
+import React from 'react';
+import { Query } from 'react-apollo';
+import gql from 'graphql-tag';
+import { LoadingMessage, ErrorMessage } from '@messages/default-messages';
+
+var lastValue = null;
+
+const getVersion = gql`
+  {
```
Comment (cbarber) [inline app/javascript/components/get-version.jsx:9]: Nit: I think specifying a name for the graphql query will provide a name in the apollo devs tools.

File: app/javascript/components/get-version.jsx [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    schemaSha
+  }
+`;
+
+export const Version = () => (
+  <Query query={getVersion}>
+    {({ loading, error, data }) => {
+      if (loading) {
+        return <LoadingMessage />;
```
Comment (cbarber) [inline app/javascript/components/get-version.jsx:18]: I think we should probably return `null` for both loading and error. No reason for this component to create dom elements.

File: app/javascript/components/get-version.jsx [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+      }
+
+      if (error) {
+        return <ErrorMessage />;
+      }
+
+      var newValue = data.schemaSha;
```
Comment (cbarber) [inline app/javascript/components/get-version.jsx:25]: Nit: this could be a const

File: app/javascript/components/get-version.jsx [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+var lastValue = null;
+
+const getVersion = gql`
+  {
+    schemaSha
+  }
+`;
+
+export const Version = () => (
+  <Query query={getVersion}>
```
Comment (cbarber) [inline app/javascript/components/get-version.jsx:15]: Does this `Query` refresh the page on a stop server/edit HEROKU_SLUG_COMMIT/start server?

File: app/javascript/components/get-version.jsx (hunk 2 above)
Comment (tmsavage) [reply app/javascript/components/get-version.jsx:30]: Yeah it was meant to be `window.location.reload()`. object is not correct.

**Review by fallynlogan — COMMENTED** [review]
Why would we want to compare the new and old SHA's?

**Comment by tmsavage** [comment]
> Why would we want to compare the new and old SHA's?

The purpose of comparing the current SHA on a client with a new SHA sent from the server is to identify if the application version on a client is out of date. Making this comparison would trigger a refresh on the client with the new received SHA, showing the most up to date version of helios.

File: app/javascript/components/get-version.jsx (hunk 6 above)
Comment (chewnoill) [reply app/javascript/components/get-version.jsx:14]: This query will only trigger a refresh after mounting

File: app/javascript/components/get-version.jsx [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+const getVersion = gql`
+  {
+    schemaSha
+  }
+`;
+
+let lastValue = null;
+
+export const Version = () => (
+  <Query query={getVersion}>
```
Comment (tucker-weed) [inline app/javascript/components/get-version.jsx:14]: Is there a reason lastValue is defined outside of the Version component? Does it make it such that new renders of the Version component don't reset lastValue?

Comment (tmsavage) [reply app/javascript/components/get-version.jsx:14]: Yes, exactly. This is a global variable, which is outside the scope of the Version component. The reason `lastValue` is a global variable is to have easy access to the current SHA value on the client. It is initialized as `null`, but is reset to the current SHA after the first instance of Version.

File: app/javascript/components/get-version.jsx [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    {({ loading, error, data }) => {
+      if (loading || error) {
+        return null;
+      }
+ 
+      const newValue = data.schemaSha;
+      if (lastValue == null) {
+        lastValue = newValue;
+        console.log(lastValue);
+        console.log(newValue);
```
Comment (tmsavage) [inline app/javascript/components/get-version.jsx:24]: `console.log()` for testing purposes.

**Review by tmsavage — COMMENTED** [review]
fixed formatting

File: app/javascript/components/get-version.jsx (hunk 3 above)
Comment (chewnoill) [reply app/javascript/components/get-version.jsx:6]: +1 this is best practice

https://www.apollographql.com/docs/react/data/operation-best-practices/#name-all-operations

File: app/javascript/components/get-version.jsx [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+const getVersion = gql`
+  {
+    schemaSha
+  }
+`;
+
+let lastValue = null;
+
+export const Version = () => (
```
Comment (cbarber) [inline app/javascript/components/get-version.jsx:13]: I'm curious if we could call the ApolloClient directly with `subscribe` to skip React entirely for our reload logic:
https://www.apollographql.com/docs/react/api/core/ApolloClient/#ApolloClient.subscribe

File: app/graphql/types/subscription_type.rb [hunk 10]
```diff
@@ -14,4 +14,8 @@ def weather_published(latitude:, longitude:); end
   field :announcement_published, Types::AnnouncementType, description: "An announcement was published"
 
   def announcement_published; end
+
+  field :schema_sha, String, description: "Updated SHA value"
+
+  def schema_sha; end
```
[the remark this answers is not in the document]
Comment (tmsavage) [reply app/graphql/types/subscription_type.rb:20]: Good point, I'll change that.

File: .circleci/config.yml [hunk 11]
```diff
@@ -57,6 +57,12 @@ jobs:
       - run: bundle exec rake db:create
       - run: bundle exec rake db:schema:load
 
+      - run:
+          name: run diff on schema.json
+          command: |
+            bundle exec rake graphql:schema:dump
+            git diff --quiet schema.graphql
+
```
Comment (chewnoill) [inline .circleci/config.yml:65]: be careful with rebasing, looks like this got pulled into your branch by mistake.  Do you know what happened here?

File: app/javascript/components/get-version.jsx (hunk 3 above)
Comment (chewnoill) [reply app/javascript/components/get-version.jsx:6]: can you use this same query as a query AND a subscription? I thought you needed to name subscriptions specifically.

File: app/javascript/components/get-version.jsx (hunk 8 above)
Comment (chewnoill) [reply app/javascript/components/get-version.jsx:24]: lets make sure not to merge these

File: .circleci/config.yml (hunk 11 above)
Comment (tmsavage) [reply .circleci/config.yml:65]: Hm, I was trying to fix my branch which diverged into two separate ones. My terminal stated it had '2 and 17 different commits each'. This is probably what was causing some errors while rebasing.

**Pull request closed without merging**